# A Multimodal Object-level Contrast Learning Method for Cancer Survival Risk Prediction

This repository is the official implementation of A Multimodal Object-level Contrast Learning Method for Cancer Survival Risk Prediction

![](figs/model-mm.png)

## ðŸš€ 1. Requirements

To install requirements:

```bash
pip install -r requirements.txt
```

## ðŸš€ 2. Download Data and Pretrained Models

1. **Downloading TCGA Data**

   All of our mutimodal data are download from public dataset repository [TCGA](https://portal.gdc.cancer.gov/repository).
2. **Preprocessing Whole Slide Images**

   We use the publicaly available WSIs-analysis toolbox [CLAM](https://github.com/mahmoodlab/CLAM.) to preprocess WSIs. All WSIs are tiled to nonoverlapping 256x256 patches at the 20x magnification and  the patches with large background area are deleted. Then a pretrained truncated ResNet50 is used to encode raw image patches into 1024-dim feature vector.

   Folder structure:

   ```
   .
   â”œâ”€â”€ LUAD
   â”‚Â Â  â””â”€â”€ features
   â”‚Â Â   Â Â  â”œâ”€â”€ pt_files
   |			â”œâ”€â”€ slide1.pt
   |	 		â”œâ”€â”€ slide2.pt
   â”‚Â Â   Â Â   Â Â  â”œâ”€â”€ ...
   â”‚Â Â   Â Â  â””â”€â”€ h5_files
   |			â”œâ”€â”€ slide1.h5
   |	 		â”œâ”€â”€ slide2.h5
   â”‚Â Â   Â Â   Â Â  â”œâ”€â”€ ...
   â”œâ”€â”€ KIRC
   â”‚Â Â  â””â”€â”€ features
   â”‚Â Â   Â Â  â”œâ”€â”€ pt_files
   |			â”œâ”€â”€ slide1.pt
   |	 		â”œâ”€â”€ slide2.pt
   â”‚Â Â   Â Â   Â Â  â”œâ”€â”€ ...
   â”‚Â Â   Â Â  â””â”€â”€ h5_files
   |			â”œâ”€â”€ slide1.h5
   |	 		â”œâ”€â”€ slide2.h5
   â”‚Â Â   Â Â   Â Â  â”œâ”€â”€ ...
   â”œâ”€â”€ ...
   ```
3. **Preprocessing RNA-Seq Data**

   We collate the patients' RNA-Seq data downloaded from the website into expression matrix and screen out cancer-related genes.

   Folder structure:

   ```bash
   .
   â”œâ”€â”€ LUAD
   â”‚Â Â  â”œâ”€â”€ 5fold-rna
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ splits_0.csv
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ splits_1.csv
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ...
   â”‚Â Â  â”œâ”€â”€ raw_data
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ clinical.csv
   â”‚Â Â  â”‚Â Â  â””â”€â”€ rnaseq.csv
   â”‚Â Â  â””â”€â”€ wsi_rna_ids.csv
   â”œâ”€â”€ KIRC
   â”‚Â Â  â”œâ”€â”€ 5fold-rna
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ splits_0.csv
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ splits_1.csv
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ...
   â”‚Â Â  â”œâ”€â”€ raw_data
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ clinical.csv
   â”‚Â Â  â”‚Â Â  â”œâ”€â”€ rnaseq.csv
   â”‚Â Â  â”‚Â Â  â””â”€â”€ wsi_rnaseq.txt
   â”‚Â Â  â””â”€â”€ wsi_rna_ids.csv
   â”œâ”€â”€ ...
   â”œâ”€â”€ mrna.csv
   â””â”€â”€ signatures.csv
   ```

## ðŸš€ 3. Training

\<cancer type\> can be LUAD or KIRC

**SNN**

```bash
CUDA_VISIBLE_DEVICES=0 python train.py\
    --dataset_dir <path to datasets dir>\
    --results_dir <path to results dir>\
    --study <cancer type>\
    --data_mode omic\
    --model_type snn\
    --model_size_omic small\
    --target_gene signatures_rnaseq\
    --reg_type omic\
    --bag_loss contrast\
    --gc 128\
    --lr 0.0002\
    --n_classes 1 
```

**AMIL**

```bash
CUDA_VISIBLE_DEVICES=0 python train.py\
    --data_dir <path to WSI features dir>\
    --dataset_dir <path to datasets dir>\
    --results_dir <path to results dir>\
    --study <cancer type>\
    --data_mode path\
    --model_type amil\
    --bag_loss contrast\
    --gc 128\
    --lr 0.0002\
    --n_classes 1
```

## ðŸš€ 4. Evaluation

**SNN**

```bash
CUDA_VISIBLE_DEVICES=0 python eval.py\
    --weights <path to model weights>\
    --fold <fold number to eval>\
    --save_dir <dir path to save results>\
    --dataset_dir <path to datasets dir>\
    --study <cancer type>\
    --data_mode omic\
    --model_type snn\
    --model_size_omic small\
    --target_gene signatures_rnaseq\
    --n_classes 1
```

**AMIL**

```bash
CUDA_VISIBLE_DEVICES=0 python eval.py\
 	--weights <path to model weights>\
    --fold <fold number to eval>\
    --save_dir <dir path to save results>\
    --data_dir <path to WSI features dir>\
    --dataset_dir <path to datasets dir>\
    --study <cancer type>\
    --data_mode path\
    --model_type amil\
    --n_classes 1
```

**late fusion**

```
python late_fusion.py\
    --wsi_dir <dir to wsi results>\
    --rna_dir <dir to genomic results>\
    --muti_dir <dir to mutimodal output>
```

## ðŸš€ 5. Visualization

We also provide our visualizetion code to explain models.

![](figures/interpreability.png)

```
python draw_heatmap.py\
    --id <wsi id>\
    --weights <wsi model weights>\
    --wsi_dir <wsi raw data dir>\
    --h5_dir <wsi h5 files dir>
```
